### PR TITLE
Fix Address Sanitizer reported error and Thread Sanitizer reported data race

### PR DESCRIPTION
**Pull request by tantei3**
Heap use after free error reported by Address Sanitizer in case of audio init failure. 

Thread sanitizer reported data race for the variable `in->byte_level_seeks` which is written without acquiring the lock inside `update_bytes_read` and simultaneously being read in `demux_get_reader_state` with lock acquired.

### CODE REVIEW

File: player/audio.c [hunk 1]
```diff
@@ -856,6 +856,9 @@ void fill_audio_out_buffers(struct MPContext *mpctx)
     if (ao_c->filter->ao_needs_update)
         reinit_audio_filters_and_output(mpctx);
 
+    if (!mpctx->ao_chain)
```
Comment (CounterPillow) [inline player/audio.c:859]: This could be `if(!ao_c)` again for consistency with the similar statement above, I believe.

**Review by CounterPillow — APPROVED** [review]
Looks good to me, the concurrency issue seems to me to be fairly inconsequential though (at least on architectures where writes to 64-bit integers are atomic and not like, two 32-bit writes), at worst the reader (a debug property) would get outdated data from what I can see, but it's still proper to have it done within the mutex.

Thank you!

File: player/audio.c (hunk 1 above)
Comment (tantei3) [reply player/audio.c:859]: The variable `ao_c` is a a local stack variable inside the `fill_audio_out_buffers` and inside the `reinit_audio_filters_and_output`, `mpctx->ao_chain` is freed and set to `NULL`, and hence the variable `ao_c` contains a reference to a freed memory location in this case, hence we can't check that, and have to check `mpctx->ao_chain` I think.

Comment (Dudemanguy) [reply player/audio.c:859]: How about just modifying `reinit_audio_filters_and_output` to just return an error code instead and then returning from here if there's an error? The function itself is already basically has error checking built into it. It's just void currently.

Comment (tantei3) [reply player/audio.c:859]: @Dudemanguy that sounds reasonable, if I understood what you're suggesting, `reinit_audio_filters_and_output` should be modified to return `-1` if ever `init_error` label is reached and in all other places return `0` and check in the above function call right?

It seems like `reinit_audio_filters_and_output` is only called in this one place inside`fill_audio_out_buffers` in the whole codebase so shouldn't be a problem modifying. But it does seem like a bigger change in terms of number of lines modified metric compared to the current simple if check though.

Comment (Dudemanguy) [reply player/audio.c:859]: >reinit_audio_filters_and_output should be modified to return -1 if ever init_error label is reached and in all other places return 0 and check in the above function call right?

Yup.

>But it does seem like a bigger change in terms of number of lines modified metric compared to the current simple if check though.

Clarity is more important imo. What you're really doing here is just bailing out if `reinit_audio_filters_and_output` fails for some reason. The `if (!mpctx->ao_chain)` looks really random compared to something like `if (reinit_audio_filters_output)` where the later makes it clear what's happening.

Comment (tantei3) [reply player/audio.c]: @Dudemanguy have made the change as per your suggestion, thanks

Comment (Dudemanguy) [reply player/audio.c]: You can squash the two audio commits together. It's part of the same change.

**Comment by Dudemanguy** [comment]
Looks good. Thanks!

**Pull request merged**